remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/matchers/UserAttributeMatcher.kt

```diff
@@ -0,0 +1,79 @@
+                return matchingAttribute.matches(userBrowserProperties.appTheme().toString())
+            }
+            is MatchingAttribute.Bookmarks -> {
+                if (matchingAttribute == MatchingAttribute.Bookmarks()) return Result.Fail
+                if (matchingAttribute.value != MATCHING_ATTR_INT_DEFAULT_VALUE) {
+                    return (matchingAttribute as IntMatchingAttribute).matches(userBrowserProperties.bookmarks().toInt())
+                }
+                return (matchingAttribute as RangeIntMatchingAttribute).matches(userBrowserProperties.bookmarks().toInt())
+            }
+            is MatchingAttribute.DaysSinceInstalled -> {
+                if (matchingAttribute == MatchingAttribute.DaysSinceInstalled()) return Result.Fail
+
+                if (matchingAttribute.value != MATCHING_ATTR_INT_DEFAULT_VALUE) {
+                    return (matchingAttribute as IntMatchingAttribute).matches(userBrowserProperties.daysSinceInstalled().toInt())
+                }
+                return (matchingAttribute as RangeIntMatchingAttribute).matches(userBrowserProperties.daysSinceInstalled().toInt())
+            }
+            is MatchingAttribute.DaysUsedSince -> {
+                val daysUsedSince = userBrowserProperties.daysUsedSince(matchingAttribute.since)
+                return matchingAttribute.matches(daysUsedSince.toInt())
+            }
+            is MatchingAttribute.DefaultBrowser -> {
+                return matchingAttribute.matches(userBrowserProperties.defaultBrowser())
+            }
+            is MatchingAttribute.EmailEnabled -> {
+                return matchingAttribute.matches(userBrowserProperties.emailEnabled())
+            }
+            is MatchingAttribute.Favorites -> {
+                if (matchingAttribute == MatchingAttribute.Favorites()) return Result.Fail
+                if (matchingAttribute.value != MATCHING_ATTR_INT_DEFAULT_VALUE) {
+                    return (matchingAttribute as IntMatchingAttribute).matches(userBrowserProperties.favorites().toInt())
+                }
+                return (matchingAttribute as RangeIntMatchingAttribute).matches(userBrowserProperties.favorites().toInt())
+            }
+            is MatchingAttribute.SearchCount -> {
+                if (matchingAttribute == MatchingAttribute.SearchCount()) return Result.Fail
+                if (matchingAttribute.value != MATCHING_ATTR_INT_DEFAULT_VALUE) {
+                    return (matchingAttribute as IntMatchingAttribute).matches(userBrowserProperties.searchCount().toInt())
+                }
+                return (matchingAttribute as RangeIntMatchingAttribute).matches(userBrowserProperties.searchCount().toInt())
+            }
+            is MatchingAttribute.WidgetAdded -> {
+                return matchingAttribute.matches(userBrowserProperties.widgetAdded())
+            }
+            else -> throw IllegalArgumentException("Invalid matcher for $matchingAttribute")
```

Comment:
Why do we want to throw here. The function already returns a result, which is kind of contradictory to throwing on error.
Why not returning `Result.Fail` (?)

Reply:
return `Result?`instead of Exception